### DISCUSSION TITLE
[12/13] 박승철, A Fast Post-Training Pruning Framework for Transformers

### DISCUSSION THREAD

**Discussion by snudm-starlab** [sjquan/2022-Study#36] [Seminars]
12/13 세미나 질문 답변입니다.

**Comment by snudm-starlab** [comment]
Task에 따라 pruning이 잘되는 layer의 위치가 다르다고 하셨는데 (p14), pruning이 잘된다의 의미는 해당 layer만 pruning을 했을때 pruning 비율 대비 성능저하가 적은것을 의미하나요?

**Reply by snudm-starlab** [reply]
네, 맞습니다. 비슷한 경향은 양자화에서 비트 수의 분포를 결정할 때에도 나타나는 것으로 알고 있습니다. 또한, FFN layer에 mask parameter를 도입해서 레이어 별로 중요도가 낮은 dimension을 제거하는 방식을 해보아도 일반적으로 언급했던 경향성들대로 중요도가 측정되는 것을 확인할 수 있습니다.

**Comment by snudm-starlab** [comment]
Residual connection이 있는 경우 pruning을 할 때는 residual connection이 이어진 부분을 모두 pruning 하나요?

**Reply by snudm-starlab** [reply]
Transformer의 경우 Layernorm(x+sublayer(x)) 꼴로 볼 수 있는데, 일반적으로 sublayer의 output 차원이 변하지 않는 프루닝을 많이 하게 됩니다. 예를 들어 FFN 레이어는 보통 768 -> 3072 -> 768 순으로 차원이 변하는데, 3072 차원을 프루닝하면 output 차원은 변하지 않고, attention head를 프루닝하는 경우도 비슷하게 동작합니다. (output projection 의 차원이 변하지 않음) 그래서 residual connection을 pruning하는 것은 필요하지 않는 것이 일반적입니다.
다만, 만약 sublayer 함수의 output의 차원이 변하게 된다면, x+sublayer(x)의 연산이 가능하게 하기 위해 residual connection도 같이 pruning을 할 수 있도록 설계하여야 할 것입니다.

**Comment by snudm-starlab** [comment]
Quantization과 Pruning의 효율성에 대한 비교 연구가 있을까요?

**Reply by snudm-starlab** [reply]
사실 이제까지 두 기법의 성능을 직접 비교한 논문을 본 적은 없고, 일반적으로 compatible한 기법으로써 함께 쓰일 수 있다고 주장하는 논문들을 많이 봤던 것 같습니다. 
(e.g. Structured Pruning of Large Language Models, EMNLP 2020)

**Comment by snudm-starlab** [comment]
Large Model 에서 Structured Pruning 을 하는 것이 단순히 결과적으로 동일한 Small NN 을 만드는 것과는 차이가 있다는 말씀을 해주셨는데요, 저는 Quantization 을 자주하는 입장으로서 말씀해주신 현상이 Task Specific Fine-Tuned 로 init된 모델을 QAT 하는 것이 Scratch 부터 QAT 하는 것 보다 월등한 성능을 내는 현상 (특히 BERT 와 같은 Pre-Training -> Fine-Tuning scheme 에서) 과 어찌보면 연관될 수 있지 않을까라는 생각이 들었습니다. (Slide 18p Exploit the pre-trained model's knowledge 한다는 측면에서..) 이렇게 Pruning 과 Quantization 에서 나타나는 현상을 함께 생각해보는 것도 참 흥미롭다는 생각이 드는데, 제가 말씀드린 연관성에 대해 간단하게라도 어떻게 생각하시는지 여쭤보고 싶고, 또한 약간은 추상적인 질문이지만 (상대적으로) 덜 필요한 부분을 제거하는 Pruning 과 모델의 Parameter 혹은 Activation 을 적은 Bit 로 represent 하는 Quantization 은 근본적으로 어떻게 접근을 다르게 하면 좋을지에 대한 승철님의 간단한 인사이트(?) 를 여쭤보고 싶습니다.

**Reply by snudm-starlab** [reply]
우선, 발표를 유심히 들어주시고, 좋은 질문을 해주셔서 감사합니다. 첫 번째 질문부터 말씀드리자면, 제가 생각했을 때 BERT에서 fine-tuned 모델을 quantize 했을 때 성능이 더 높은 이유는 아마도 사용된 데이터의 차이라고 생각합니다. BERT의 경우에는 task와 관계없이 일반적으로 사용될 수 있는 knowledge를 매우 큰 corpus를 통해서 학습한 후에 downstream task에 대해 fine-tuned되는 것이고, scratch부터 학습된 모델은 적은 수의 down stream data만 보고 학습되는 것이기 때문에, 이러한 데이터의 접근성 차이에 의해 성능이 더 높게 나올수 있다고 생각합니다. 아마도 down-stream data만으로는 큰 데이터로부터 얻을 수 있는 general한 knowledge를 얻기 힘들기 때문일 것입니다. 그리고 만약 scratch부터 학습하는 작은 모델이 pre-training에 해당하는 과정까지도 모두 동일하게 학습한 후에 똑같이 fine-tuning하고 QAT를 진행한다면 아마도 Large 모델을 pruning 한 것과 비슷한 성능을 내지 않을까 생각됩니다. 해당 내용은 pruning/ quantization과 관계없이 일반적으로 통용될 수 있는 내용이기 때문에 두 기법에서 모두 비슷한 경향을 보이지 않았나 생각합니다.

**Reply by snudm-starlab** [reply]
다음으로, Structured pruning 을 진행하게 되면 일반적으로 어떠한 공간의 차원이 줄어들게 됩니다. 예를 들어 Weight matrix (n x m)을 x (1 x m)와 곱해서 xW를 계산한다는 것은 이는 x를 m에서 n 차원 공간으로 보내는 것입니다. 여기서 W의 row를 k개 pruning 한다면, x를 m에서 (n-k)차원으로 보내게 되고, x가 존재하는 공간의 차원이 줄어들게 되는 것입니다. 그리하여 LRA와 비슷한 논지들이 많이 나오게 되고, 해당 내용들도 발표자료에 포함하게 된 것입니다. (또 다른 예로 attention score를 계산하는 과정이 xW_Q W_K ^T x^T 꼴로 나타낼 수 있는데 이는 MF가 이미 이루어진 꼴과 같습니다.)
 그러므로 structured pruning을 하게 되면 공간의 차원이 줄어들고, 좀 더 자세히는 matrix의 rank가 줄어들기 때문에 이를 잘 고려해야한다고 생각됩니다.  특히, non-linearity는 neural net의 복잡도를 증가시켜주는 핵심적인 부분인데, 이 것이 적용되는 부분들에 대해서 차원을 줄이는 것은 타격이 클 수도 있습니다.
이와 달리 quantization은 우리가 수학적으로 표현해놓은 것을 컴퓨터로 구현하고 연산하는 과정에서의 precision을 결정하는 것으로써 이 것이 수식의 의미를 변화시키지는 않고, 오차가 커지는 개념으로 생각됩니다. 우리는 64bit의 double 대신 32bit로도 충분히 원하는 결과를 얻을 수 있음을 경험적으로 알게되어 32bit를 Full precision으로 주로 사용해왔고, 여기서 16bit 까지는 성능 하락이 크지 않게 많이 사용되고 그 이하로 가게 되면 오차를 잡기 위해 twin 구조를 쓰는 등 다양한 기법들이 쓰이는 것 같습니다. 이 부분은 제가 논문을 많이 읽어본 것은 아니라서 간단하게만 쓰겠습니다.
결론적으로, structured pruning은 모델의 구조적인 redundant함을 줄이기 위함이고, quantization은 계산의 표현에 대한 redundant함을 줄이기 위한 것으로써, 둘은 어느 정도 compatible한 관계가 아닐까 생각이 듭니다.

**Comment by snudm-starlab** [comment]
norm based pruning이 좋지 않는 이유는 뭔가요?

**Reply by snudm-starlab** [reply]
Norm-based pruning이라고 함은 아마도 CNN 쪽에서 NAS (Neural Architecture Search)기반의 기법들을 사용하지 않은 기법들을의미하는 것 같은데 사실 구분이 살짝 애매한 면이 있는 것 같습니다. 예를 들어 attention head들에 mask 파라미터를 부여하고, 이들 중에 중요한 parameter를 찾는 것 자체가 NAS 에서의 문제 세팅과 비슷해지기 때문입니다. 다만, 여기서 NAS 기반의 기법들을 사용한다고 한다면, mask의 큰 조합들에 대한 search space를 효과적으로 줄이는 방법론들과 mask의 조합이 변함에 따라 변하는 모델의 성능 예측을 효과적으로 추정하는 방법론들이 들어가기 때문에 조금 더 좋다고 할 수 있을 것 같습니다.

**Reply by snudm-starlab** [reply]
추가적으로 ViT 등 Transformer 기반의 structured pruning 논문의 경우에는 Group lasso를 기반으로 압축하는 기법들이 최근에도 발표되고 있으며, 대표적으로 발표때 소개해드린 Unified Visual Transformer Compression (ICLR 2022) 논문이 있습니다.

**Comment by snudm-starlab** [comment]
질문해주신 내용들과 더불어 설문에서 "RETHINKING THE VALUE OF NETWORK PRUNING, ICLR 2019" 논문을 추받아 읽어보았는데 제가 진행한 실험 결과의 내용과 아주 유사한 내용이 담겨있어 공유 드립니다.
해당 논문은 pruning으로 찾은 neural network의 파라미터들을 초기화 하여 학습하였을 때 성능의 변화를 관측합니다. 즉, pruning으로 찾은 아키텍쳐와 그 파라미터 조합을 구분하여 둘 중에 어떤 것이 의미있는가를 보여주는데, structured pruning에 한해서는 architecture를 잘 찾는 효과가 더 크며, 파라미터 조합 자체는 scratch부터 학습한 것이 더 좋았다고 합니다.

**Reply by snudm-starlab** [reply]
논문의 링크는 아래와 같으며, 관심 있으신 분들은 읽어보시면 좋을 것 같습니다. https://arxiv.org/pdf/1810.05270.pdf
